backends/smt2/smtio.py

```diff
@@ -31,16 +31,16 @@
 # does not run out of stack frames when parsing large expressions
 if os.name == "posix":
     smtio_reclimit = 64 * 1024
-    smtio_stacksize = 128 * 1024 * 1024
-    smtio_stacklimit = resource.RLIM_INFINITY
-    if os.uname().sysname == "Darwin":
-        # MacOS has rather conservative stack limits
-        smtio_stacksize = 16 * 1024 * 1024
-        smtio_stacklimit = resource.getrlimit(resource.RLIMIT_STACK)[1]
     if sys.getrecursionlimit() < smtio_reclimit:
         sys.setrecursionlimit(smtio_reclimit)
-    if resource.getrlimit(resource.RLIMIT_STACK)[0] < smtio_stacksize:
-        resource.setrlimit(resource.RLIMIT_STACK, (smtio_stacksize, smtio_stacklimit))
+
+    smtio_stacksize = 128 * 1024 * 1024
+    current_rlimit_stack = resource.getrlimit(resource.RLIMIT_STACK)
+    if current_rlimit_stack[0] != resource.RLIM_INFINITY:
+        if current_rlimit_stack[1] != resource.RLIM_INFINITY:
+            smtio_stacksize = min(smtio_stacksize, smtio_stacklimit)
+        if current_rlimit_stack[0] < smtio_stacksize:
+            resource.setrlimit(resource.RLIMIT_STACK, (smtio_stacksize, current_rlimit_stack[1]))
```

Comment:
this does not work on macOS. 
Should not it be:
resource.setrlimit(resource.RLIMIT_STACK, (current_rlimit_stack[0], smtio_stacksize))

Reply:
Hmm.. I don't think so. The 2nd tuple element is the hard limit. You get a `ValueError`?

Reply:
setrlimit(...) throws a ValueError for anything over 16MB of MacOS. One can run successfully run "ulimit -s $(ulimit -Hs)" though, which puts the stack at 64MB (ulimit -a shows the expected limit). I am not sure where this falls apart, I have not had a chance to look at Python interpreter source so see if this exception is some artificial limit.

Reply:
I poked at this for a bit and using getrlimit(...)/setrlimit(...) in a C++ program works as expected. You can set the soft limit to the max and afterwards allocate a std::array on the next stack frame with a size close to the hard limit. The CPython does pretty much the same thing (https://github.com/python/cpython/blob/d2b11af91560eaaeb499f702f4b0f244ec756280/Modules/resource.c#L216), but for some reason seems to run into an EPERM errno. Not sure what's going on here.

Reply:
Well others as well have same issue on  macOS with python3 https://github.com/sagemath/cysignals/issues/71  but there is no actual fix for this issue there.

Reply:
I tried this with py2.7 on MacOS and it works as expected. Thanks Apple :/

Reply:
@mmicko That sage issue is about the message for EINVAL, but @arjenroodselaar says in our case we get EPERM.

Anyway, I've now pushed another commit that limits the stack size we attempt to set to 16 MB on Darwin. Grrr. This is not very satisfying..